StackExchange.Redis/StackExchange/Redis/ConnectionMultiplexer.cs

```diff
@@ -2058,37 +2069,60 @@ internal T ExecuteSyncImpl<T>(Message message, ResultProcessor<T> processor, Ser
             }
         }
 
+        private static void AppendTimeoutHelpLinkUrl(StringBuilder sb)
+        {
+            sb.Append(" Please take a look at this article for some tips for investigating timeouts: ");
+            sb.Append(timeoutHelpLink);
+            sb.Append(".");
+        }
+
 #if !CORE_CLR
-        private static string GetSystemCpuPercent()
+        /// <summary>
+        /// Gathers system stats and appends detailed info about possible causes of timeouts to the error message StringBuilder
+        /// </summary>
+        private static void AddTimeoutRootCauseIfPossible(StringBuilder sb, ServerEndPoint server, out ThreadPoolStats iocp, out ThreadPoolStats worker, out string localCpuPercent)
         {
+            localCpuPercent = "unavalable";
+            ThreadPoolStats.Gather(out iocp, out worker);
+
             float systemCPU;
-            if (PerfCounterHelper.TryGetSystemCPU(out systemCPU))
+            bool gotCpuUsageSuccessfully = PerfCounterHelper.TryGetSystemCPU(out systemCPU);
+
+            if (server.ConnectionState != PhysicalBridge.State.ConnectedEstablished)
             {
-                return Math.Round(systemCPU, 2) + "%";
+                sb.Append($" The connection to the server was lost and was not restored in the timeout specified.");
+                return; // don't add IOCP and CPU details to the error message because connection management trumps all other causes of timeouts.
+            }
+
+            bool detailsAdded = false;
+            bool timeoutHelpLinkAdded = false;
+            if (iocp.Busy > iocp.Min || worker.Busy > worker.Min)
+            {
+                detailsAdded = true;
+                timeoutHelpLinkAdded = true;
+                sb.Append($" The number of busy IOCP or WORKER threads in the ThreadPool is greater than the 'Min' setting, which could easily be the cause of this timeout.  See https://github.com/StackExchange/StackExchange.Redis/blob/master/Docs/Timeouts.md#are-you-seeing-high-number-of-busyio-or-busyworker-threads-in-the-timeout-exception for details on how ThreadPool Growth Throttling can affect performance.");
+            }
+            
+            if (gotCpuUsageSuccessfully)
+            {
+                localCpuPercent = Math.Round(systemCPU, 2) + "%";
+                if (systemCPU > 85)
+                {
+                    if (detailsAdded)

```

Comment:
I'd simplify this - the "Additionally, " isn't really needed and the code is much simpler without the differentiation.